torch/distributed/fsdp/_state_dict_utils.py

```diff
@@ -599,55 +590,63 @@ def _sharded_pre_load_state_dict_hook(
-            if local_tensor.device.type != pg_device.type:
-                local_tensor = local_tensor.to(pg_device)
-            num_padding = chunk_size - local_tensor.numel()
-            if num_padding > 0:
-                local_tensor = F.pad(local_tensor, [0, num_padding])
-        else:
-            local_tensor = torch.zeros(chunk_size, dtype=param.dtype, device=device)
-        tensor = torch.empty(
-            chunk_size * fsdp_state.world_size, dtype=local_tensor.dtype, device=device
-        )
-        if local_tensor.is_cpu:
-            tensor_list = list(
-                torch.chunk(tensor, dist.get_world_size(fsdp_state.process_group))
+        if not fsdp_state._state_dict_config.use_dtensor:
+            # All-gather the param (ShardedTensor)
+            param, shards = _ext_pre_load_state_dict_transform(param)
+
+            assert len(shards) < 2, (
+                "Expects 0 or 1 shard per rank "
+                f"but got {len(shards)} shards on rank {fsdp_state.rank}."
+            )
+            param_numel = param.size().numel()
+            dim_0_size = param.size()[0]
+            chunk_size = (
+                math.ceil(dim_0_size / fsdp_state.world_size)
+                * param_numel
+                // dim_0_size
+            )
+            if len(shards) == 1:
+                local_tensor = shards[0].tensor.flatten()
+                pg_device = _get_pg_default_device(fsdp_state.process_group)
+                if local_tensor.device.type != pg_device.type:
+                    local_tensor = local_tensor.to(pg_device)
+                num_padding = chunk_size - local_tensor.numel()
+                if num_padding > 0:
+                    local_tensor = F.pad(local_tensor, [0, num_padding])
+            else:
+                local_tensor = torch.zeros(chunk_size, dtype=param.dtype, device=device)
+            tensor = torch.empty(
+                chunk_size * fsdp_state.world_size,
+                dtype=local_tensor.dtype,
+                device=device,
             )
-            dist.all_gather(tensor_list, local_tensor, group=fsdp_state.process_group)
+            if local_tensor.is_cpu:
+                tensor_list = list(
+                    torch.chunk(tensor, dist.get_world_size(fsdp_state.process_group))
+                )
+                dist.all_gather(
+                    tensor_list, local_tensor, group=fsdp_state.process_group
+                )
+            else:
+                dist.all_gather_into_tensor(
+                    tensor, local_tensor, group=fsdp_state.process_group
+                )
+            tensor = tensor.narrow(0, 0, param_numel).reshape(param.size())
+            state_dict[fqn_from_global_root] = tensor
         else:
-            dist.all_gather_into_tensor(
-                tensor, local_tensor, group=fsdp_state.process_group
+            # TODO: make use_dtensor=True work with offload_to_cpu=True.
+            param = param.redistribute(
```

Comment:
If param is in CPU, will this function still work? We need to consider parameters in `state_dict` are loaded to CPU first before calling `load_state_dict()`.

Reply:
> If param is in CPU, will this function still work? We need to consider parameters in `state_dict` are loaded to CPU first before calling `load_state_dict()`.

Good catch. Thanks for raising this! Updated to handle this case and included test cases to cover this.